### PR TITLE
Windows: expand nested environment variables

### PR DESCRIPTION
**Pull request by erw7**
Fixed an issue where  Test_WindowsHome() fails on oldtest.

### CODE REVIEW

File: src/nvim/testdir/test_windows_home.vim [hunk 1]
```diff
@@ -60,18 +60,18 @@ func Test_WindowsHome()
     let $USERPROFILE = 'unused'
     let $HOMEDRIVE = 'C:'
     let $HOMEPATH = '\foobar'
-    let $HOME = ''  " Force recomputing "homedir"
+    unlet $HOME  " Force recomputing "homedir"
```
Comment (justinmk) [inline src/nvim/testdir/test_windows_home.vim]: If plugins/tests really expect the Vim behavior perhaps we should mimic it after all. @blueyed

File: src/nvim/os/env.c [hunk 2]
```diff
@@ -429,24 +433,36 @@ void expand_env_esc(char_u *restrict srcp,
...
           }
         }
 
-#if defined(UNIX)
+#if defined(UNIX) || defined(WIN32)
+# ifdef UNIX
         // Verify that we have found the end of a Unix ${VAR} style variable
         if (src[1] == '{' && *tail != '}') {
+# else
+        if (*src == '%' && *tail != '%') {
```
Comment (justinmk) [inline src/nvim/os/env.c]: we could define these platform-specific start/end , though I guess it doesn't help much :

```c
#if defined(UNIX)
# define ENV_START = '%';
# define ENV_END = '%';
#else
# define ENV_START = '%';
# define ENV_END = '%';
#endif
```

File: src/nvim/testdir/test_windows_home.vim (hunk 1 above)
Comment (blueyed) [reply src/nvim/testdir/test_windows_home.vim]: Yes, plugins are an issue - it is bad if they see different behavior in subtle areas.
But for this test: this could/should be changed like this in Vim itself already - I assume it is from before `unlet` supported envs.

Comment (blueyed) [reply src/nvim/testdir/test_windows_home.vim]: @erw7 
Why is this change necessary?

File: src/nvim/os/env.c (hunk 2 above)
Comment (erw7) [reply src/nvim/os/env.c]: It seemed like a good idea, but when I examined it, I found that we could not reduce the `#ifdef`. It is rather hard to see, so it seems better to quit.

File: src/nvim/testdir/test_windows_home.vim (hunk 1 above)
Comment (erw7) [reply src/nvim/testdir/test_windows_home.vim]: This is because vim deletes environment variables with `let $FOO=''` on Windows. It seems that there is no way to create an empty environment variable other than calling the API on Windows. Therefore, I changed it because it seems that the same operation as vim is better.

Comment (blueyed) [reply src/nvim/testdir/test_windows_home.vim]: I see.  We need `unlet` to unset it, because setting it to an empty value does not unset it, right?
See also the discussion in https://github.com/pdbpp/pdbpp/pull/268 (for `exists("$EMPTY")`).

Comment (blueyed) [reply src/nvim/testdir/test_windows_home.vim]: Given your latest commit this could use the Vim version of the test already, couldn't it?

File: src/nvim/os/env.c [hunk 3]
```diff
@@ -429,24 +436,36 @@ void expand_env_esc(char_u *restrict srcp,
...
             tail++;
           }
 #endif
         *var = NUL;
         var = (char_u *)vim_getenv((char *)dst);
         mustfree = true;
-#if defined(UNIX)
+#if defined(UNIX) || defined(WIN32)
         }
 #endif
```
Comment (justinmk) [inline src/nvim/os/env.c]: `#if defined(UNIX) || defined(WIN32)` could be removed, it is always true.

File: src/nvim/os/env.c [hunk 4]
```diff
@@ -429,24 +436,36 @@ void expand_env_esc(char_u *restrict srcp,
...
+#ifdef WIN32
+                                             || (*src == '%' && *tail != '%')
+#endif
+                                             )) {  //  NOLINT(whitespace/parens)
             *var++ = *tail++;
           }
         }
 
-#if defined(UNIX)
+#if defined(UNIX) || defined(WIN32)
```
Comment (justinmk) [inline src/nvim/os/env.c]: this guard can be removed , it's always true.

**Comment by erw7** [comment]
Changes to expand Windows style environment variables in `expand_env_esc()` were unrelated to `Test_WindowsHome()`. Therefore it was separated into #10667.

File: src/nvim/testdir/test_windows_home.vim (hunk 1 above)
Comment (erw7) [reply src/nvim/testdir/test_windows_home.vim]: Yes, no test changes are needed.

**Comment by justinmk** [comment]
LGTM, is this ready?

**Comment by erw7** [comment]
> LGTM, is this ready?

@justinmk Yes, but if it is better to merge with #10679, I will merge with #10679.

**Review by blueyed — APPROVED** [review]

File: src/nvim/os/env.c [hunk 5]
```diff
@@ -119,6 +119,9 @@ int os_setenv(const char *name, const char *value, int overwrite)
   if (!overwrite && os_getenv(name) != NULL) {
     return 0;
   }
+  if (value[0] == '\0') {
+    return os_unsetenv(name);
+  }
```
Comment (justinmk) [inline src/nvim/os/env.c]: Let's do this in a separate PR, it's a rather subtle but important change. Is it needed to make some oldtests pass? Can you mention some of those oldtests in the commit message?

File: src/nvim/os/env.c [hunk 6]
```diff
@@ -300,6 +303,36 @@ void init_homedir(void)
...
   }
+
+  // Weird but true: $HOME may contain an indirect reference to another
+  // variable, esp. "%USERPROFILE%".  Happens when $USERPROFILE isn't set
+  // when $HOME is being set.
+  if (var != NULL && *var == '%') {
+    char *p;
+    const char *exp;
+
+    p = strchr(var + 1, '%');
```
Comment (justinmk) [inline src/nvim/os/env.c]:
```suggestion
    char *p = strchr(var + 1, '%');
```

File: src/nvim/os/env.c [hunk 7]
```diff
@@ -300,6 +303,36 @@ void init_homedir(void)
...
+    if (p != NULL) {
+      vim_snprintf(os_buf, (size_t)(p - var), "%s", var + 1);
+      exp = os_getenv(os_buf);
```
Comment (justinmk) [inline src/nvim/os/env.c]:
```suggestion
      const char *exp = os_getenv(os_buf);
```

File: src/nvim/os/env.c [hunk 8]
```diff
@@ -300,6 +303,36 @@ void init_homedir(void)
...
+  if (var != NULL && *var == NUL) {
+    // empty is same as not set
+    var = NULL;
+  }
+
+  // Default home dir is C:/
+  // Best assumption we can make in such a situation.
+  if (var == NULL) {
+    var = "C:/";
+  }
```
Comment (justinmk) [inline src/nvim/os/env.c]: this could be `if (var == NULL || *var == NUL)`, then the block at line 326 could be removed.

File: src/nvim/os/env.c (hunk 5 above)
Comment (blueyed) [reply src/nvim/os/env.c]: > Is it needed to make some oldtests pass?

Probably not - Vim does not want to test this behavior.
(there might be an existing one though)

Comment (erw7) [reply src/nvim/os/env.c]: Separated into #10743. 

> Is it needed to make some oldtests pass?

I checked it, but there seems to be no old tests that need this other than `Test_WindowsHome()`.

Comment (blueyed) [reply src/nvim/os/env.c]: Thanks!  I've thought `Test_WindowsHome` also would not need it really, but it uses this method.
